frontend/catalyst/tracing/contexts.py

```diff
@@ -41,6 +41,79 @@
 logger.addHandler(logging.NullHandler())
 
 
+class GradContext:
+    """This class tells us the level of nestedness of grad."""
+
+    # Conceptually:
+    #
+    #   _grad_stack : List[bool] = []
+    #
+    # is a class member variable that reflects the level of
+    # nested-ness of the grad operation. If we have the following code:
+    #
+    #   grad(fun)(x)
+    #
+    # and we are inside fun, then List _grad_stack will contain [True]
+    # if we have:
+    #
+    #  grad(grad(fun))(x)
+    #
+    # and we are tracing fun, then _grad_stack will contain [True, True]
+    #
+    # I say conceptually because you can already tell that we don't need
+    # anything besides an integer that gets incremented and decremented
+    # each time you enter a grad or jacobian operation.
+    #
+    # The reason why we started with a stack is because I think it makes more sense.
+    #
+    # You might also ask, why do we need keep track of the order of the derivative?
+    # Isn't it enough to know that we are inside the grad context or outside?
+    # It is when you are only limited to the first order derivative, but once
+    # you have a higher order derivatives, it will be important to know whether
+    # you need the derivative of the derivative...
+    #
+    # The final question to be asked here is whether we need a new stack for this
+    # context or whether we should reuse the old one.
+    # I think we should keep it simple and this is simpler.
+
+    _grad_stack: int = 0
+    # This message will be used in an assertion because it is not expected
+    # to be a user facing error ever.
+    _error_msg: str = """This is an impossible state. One cannot derive to
+derive to a negative order / integrate"""
```

Comment:
Minor: Is there a reason for this message to be a class variable? (locality)

Reply:
Ah, originally I had the assertion in two functions here.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/920/commits/1e543538dcfb28dde111faed790cd5ab58790fe5

Reply:
Ah, the other reason was to avoid having to update the test if the error message changes. See this change I had to add once the _error_msg was moved to the function:

https://github.com/PennyLaneAI/catalyst/pull/920/commits/792b0f1bb1368f7af8caf033a32079fb2a0a9ac5